File: JavaScript/001.완주하지못한선수.js
function solution(participant, completion) {
    var answer = '';
    // 오롯이 내가 풀어 본 건데... 나름 머리 열심히 굴려서 썼는데 시간 초과라고 채점 자체를 거절당했다ㅠㅠ
    var count = 0;
    for ( var i=0 ; i<participant.length ; ) {
        for ( var j=0 ; j<completion.length ; ) {
            if ( participant[i] == completion[j] ) {
                i++;
            }
            count++;
        }
        if ( count == completion.length ) {
            answer = participant[i];
            return answer;
        }
    }
    
    // 가장 일반적인 방법인 듯. 각 배열을 sort 해서 아예 같은지 비교하고, 다른 걸 answer 변수에 넣는 방식.
    participant.sort();
    completion.sort();
    
    for (var i=0; i<participant.length; i++){
        if(participant[i] !== completion[i]){
            answer = participant[i];
            return answer
        }
    }
    
    // map 을 사용해서 풀이한 좋은 코드라고 한다! map 부분 공부하면 이해할 수 있도록 주석 붙여놔야지
    const map = new Map();

    for(let i = 0; i < participant.length; i++) {
        let a = participant[i], 
            b = completion[i];

        map.set(a, (map.get(a) || 0) + 1);
        map.set(b, (map.get(b) || 0) - 1);
    }
    
    for(let [k, v] of map) {
        if(v > 0) return k;
    }

    return 'nothing';

    // 뭐지... 천잰가...?
    var solution=(participant,completion)=>participant.find(name=>!completion[name]--,completion.map(name=>completion[name]=(completion[name]|0)+1))
}

File: Python/001.완주하지못한선수.py
## JS 에서 푼 거 그대로 python 으로 구현하려고 했는데 똑같이는 안돌아가더라...
def solution(participant, completion):
    answer = ''
    
    participant.sort()
    completion.sort()
    
    for i in range(len(completion)):
        if (participant[i] != completion[i]):
            answer = participant[i];
            return answer
          
    ## 마지막에 끝까지 안맞을 경우에는 마지막 요소의 이름이 완주하지 못한 선수라는 코드도 넣어줘야 했음
    answer = participant[-1];
    return answer

##  해시함수로 구현한 거라고! 이게 가장 문제에 적합한 답이라는데 JS 는 map, python 은 해시 공부해야지
def solution(participant, completion):
    answer = ''
    temp = 0
    dic = {}
    for part in participant:
        dic[hash(part)] = part
        temp += int(hash(part))
    for com in completion:
        temp -= hash(com)
    answer = dic[temp]

    return answer

## collections 랑 counter 가 뭐죠...
import collections

def solution(participant, completion):
    answer = collections.Counter(participant) - collections.Counter(completion)
    return list(answer.keys())[0]
